### PR TITLE
Add Google APIs content

### PR DESCRIPTION
**Pull request by kevmoo**
Adds initial content towards https://github.com/flutter/website/issues/5637

### CODE REVIEW

**Comment by kevmoo** [comment]
CC @mit-mit @jonasfj @RedBrogdon

File: src/docs/development/data-and-backend/google-apis.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+Google services that you can use from Dart projects.
+
+This page describes how to use APIs that interact with end-user data by using
+Gooogle authentication.
+
+Examples of user-data APIs include
+[Calendar](https://pub.dev/documentation/googleapis/latest/calendar.v3/calendar.v3-library.html),
+[Gmail](https://pub.dev/documentation/googleapis/latest/gmail.v1/gmail.v1-library.html),
+and
+[YouTube](https://pub.dev/documentation/googleapis/latest/youtube.v3/youtube.v3-library.html).
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/google-apis.md:18]:
```suggestion
[Calendar]({{site.pub-api}}/googleapis/latest/calendar.v3/calendar.v3-library.html),
[Gmail]({{site.pub-api}}/googleapis/latest/gmail.v1/gmail.v1-library.html),
and
[YouTube]({{site.pub-api}}/googleapis/latest/youtube.v3/youtube.v3-library.html).
```

File: src/docs/development/data-and-backend/google-apis.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+APIs that require
+[service accounts](https://cloud.google.com/iam/docs/service-accounts) **_should
+not_** be used directly from a Flutter application. Doing so requires shipping
+service credentials as part of your application, which is not secure. To use
+these APIs, we recommend creating a intermediate service.
+
+<!-- TODO(kevmoo): Add link to public user guide when available. -->
+
+{{site.alert.end}}
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/google-apis.md:31]:
```suggestion
{{site.alert.info}} 
  Note: The only APIs you should use directly from your Flutter
  project are those that access user data via Google authentication.

  APIs that require
  [service accounts](https://cloud.google.com/iam/docs/service-accounts) **should
  not** be used directly from a Flutter application. Doing so requires shipping
  service credentials as part of your application, which is not secure. To use
  these APIs, we recommend creating an intermediate service.

<!-- TODO(kevmoo): Add link to public user guide when available. -->

{{site.alert.end}}
```

File: src/docs/development/data-and-backend/google-apis.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+## 1. Pick the desired API
+
+The documentation for [package:googleapis] lists each API as a separate Dart
+library – in a `name.version` format. Let's look at
+[`youtube.v3`](https://pub.dev/documentation/googleapis/latest/youtube.v3/youtube.v3-library.html)
+as an example.
+
+Each library may provide many types, but there is one _root_ class that ends in
+`Api`. For YouTube, it's
+[`YouTubeApi`](https://pub.dev/documentation/googleapis/latest/youtube.v3/YouTubeApi-class.html).
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/google-apis.md:51]:
```suggestion
[`YouTubeApi`]({{site.pub-api}}/googleapis/latest/youtube.v3/YouTubeApi-class.html).
```

File: src/docs/development/data-and-backend/google-apis.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+1. Authenticate user with the required scopes
+1. Obtain an authenticated HTTP client
+1. Create and use the desired API class
+
+## 1. Pick the desired API
+
+The documentation for [package:googleapis] lists each API as a separate Dart
+library – in a `name.version` format. Let's look at
+[`youtube.v3`](https://pub.dev/documentation/googleapis/latest/youtube.v3/youtube.v3-library.html)
+as an example.
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/google-apis.md:47]:
```suggestion
library – in a `name.version` format. For example, consider the
[`youtube.v3`]({{site.pub-api}}/googleapis/latest/youtube.v3/youtube.v3-library.html)
package.
```

File: src/docs/development/data-and-backend/google-apis.md (hunk 3 above)
Comment (sfshaza2) [inline src/docs/development/data-and-backend/google-apis.md:51]:
```suggestion
[`YouTubeApi`]({{site.pub-api}}/googleapis/latest/youtube.v3/YouTubeApi-class.html).
```

File: src/docs/development/data-and-backend/google-apis.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+`Api`. For YouTube, it's
+[`YouTubeApi`](https://pub.dev/documentation/googleapis/latest/youtube.v3/YouTubeApi-class.html).
+
+Not only is the `Api` class the one you need to instantiate – see step 3 – but
+it also exposes the scopes that represent the permissions needed to use the API.
+Look under the
+[Constants section](https://pub.dev/documentation/googleapis/latest/youtube.v3/YouTubeApi-class.html#constants)
+of the `YouTubeApi` class and you'll see the available scopes. To request access
+to simply read (but not write) an end-users YouTube data, use the
+[`youtubeReadonlyScope`](https://pub.dev/documentation/googleapis/latest/youtube.v3/YouTubeApi/youtubeReadonlyScope-constant.html)
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/google-apis.md:59]:
```suggestion
[Constants section]({{site.pub-api}}/googleapis/latest/youtube.v3/YouTubeApi-class.html#constants)
of the `YouTubeApi` class and you'll see the available scopes. To request access
to simply read (but not write) an end-users YouTube data, use the
[`youtubeReadonlyScope`]({{site.pub-api}}/googleapis/latest/youtube.v3/YouTubeApi/youtubeReadonlyScope-constant.html)
```

File: src/docs/development/data-and-backend/google-apis.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+<?code-excerpt "lib/main.dart" skip="13" take="2"?>
+```dart
+/// Provides the `YouTubeApi` class.
+import 'package:googleapis/youtube/v3.dart';
+```
+
+## 2. Authenticate the user with the required scopes.
+
+Use the [google_sign_in](https://pub.dev/packages/google_sign_in) package to
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/google-apis.md:70]:
```suggestion
Use the [google_sign_in]({{site.pub-pkg}}/google_sign_in) package to
```

File: src/docs/development/data-and-backend/google-apis.md [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+authenticate users with Google in your application. Note that you will have to
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/google-apis.md:71]:
```suggestion
authenticate users with their Google identity. You will have to
```

File: src/docs/development/data-and-backend/google-apis.md [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+configure signin for each platform you want to support.
+
+<?code-excerpt "lib/main.dart" skip="10" take="2"?>
+```dart
+/// Provides the `GoogleSignIn` class
+import 'package:google_sign_in/google_sign_in.dart';
+```
+
+When you instantiate the
+[`GoogleSignIn`](https://pub.dev/documentation/google_sign_in/latest/google_sign_in/GoogleSignIn-class.html)
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/google-apis.md:81]:
```suggestion
[`GoogleSignIn`]({{site.pub-api}}/google_sign_in/latest/google_sign_in/GoogleSignIn-class.html)
```

File: src/docs/development/data-and-backend/google-apis.md [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+<?code-excerpt "lib/main.dart" skip="35" take="3"?>
+```dart
+final _googleSignIn = GoogleSignIn(
+  scopes: <String>[YouTubeApi.youtubeReadonlyScope],
+);
+```
+
+Follow the instructions provided by
+[package:google_sign_in](https://pub.dev/packages/google_sign_in) to allow a
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/google-apis.md:92]:
```suggestion
[package:google_sign_in]({{site.pub-pkg}}/google_sign_in) to allow a
```

File: src/docs/development/data-and-backend/google-apis.md [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+user to authenticate.
+
+Once authenticated, you must obtain an authenticated HTTP client.
+
+## 3. Obtain an authenticated HTTP client.
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/google-apis.md:97]: Delete the period from a header.

File: src/docs/development/data-and-backend/google-apis.md [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+[`youtubeReadonlyScope`](https://pub.dev/documentation/googleapis/latest/youtube.v3/YouTubeApi/youtubeReadonlyScope-constant.html)
+when authenticating the user.
+
+<?code-excerpt "lib/main.dart" skip="13" take="2"?>
+```dart
+/// Provides the `YouTubeApi` class.
+import 'package:googleapis/youtube/v3.dart';
+```
+
+## 2. Authenticate the user with the required scopes.
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/google-apis.md:68]: Delete the period from the header.

File: src/docs/development/data-and-backend/google-apis.md [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+Once authenticated, you must obtain an authenticated HTTP client.
+
+## 3. Obtain an authenticated HTTP client.
+
+The
+[extension_google_sign_in_as_googleapis_auth](https://pub.dev/packages/extension_google_sign_in_as_googleapis_auth)
+package provides an
+[extension method](https://dart.dev/guides/language/extension-methods) on
+`GoogleSignIn`:
+[`authenticatedClient`](https://pub.dev/documentation/extension_google_sign_in_as_googleapis_auth/latest/extension_google_sign_in_as_googleapis_auth/GoogleApisGoogleSignInAuth/authenticatedClient.html).
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/google-apis.md:104]:
```suggestion
The
[extension_google_sign_in_as_googleapis_auth]({{site.pub-pkg}}/extension_google_sign_in_as_googleapis_auth)
package provides an
[extension method]({{site.dart-site}}/guides/language/extension-methods) on
`GoogleSignIn`:
[`authenticatedClient`]({{site.pub-api}}/extension_google_sign_in_as_googleapis_auth/latest/extension_google_sign_in_as_googleapis_auth/GoogleApisGoogleSignInAuth/authenticatedClient.html).
```

File: src/docs/development/data-and-backend/google-apis.md [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+
+<?code-excerpt "lib/main.dart" skip="7" take="1"?>
+```dart
+import 'package:extension_google_sign_in_as_googleapis_auth/extension_google_sign_in_as_googleapis_auth.dart';
+```
+
+You can listen to
+[`onCurrentUserChanged`](https://pub.dev/documentation/google_sign_in/latest/google_sign_in/GoogleSignIn/onCurrentUserChanged.html).
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/google-apis.md:112]:
```suggestion
[`onCurrentUserChanged`]({{site.pub-api}}/google_sign_in/latest/google_sign_in/GoogleSignIn/onCurrentUserChanged.html).
```

File: src/docs/development/data-and-backend/google-apis.md [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+When event value is not `null`, you can create an authenticated client.
+
+<?code-excerpt "lib/main.dart" skip="124" take="1"?>
+```dart
+var httpClient = (await _googleSignIn.authenticatedClient())!;
+```
+
+This
+[`Client`](https://pub.dev/documentation/http/latest/http/Client-class.html)
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/google-apis.md:121]:
```suggestion
[`Client`]({{site.pub-api}}/http/latest/http/Client-class.html)
```

File: src/docs/development/data-and-backend/google-apis.md [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+instance includes the nessesary credentials when invoking Google API classes.
+
+## 4. Create and use the desired API class
+
+Use it to create the desired API type and call methods, for instance:
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/google-apis.md:126]:
```suggestion
Use the API to create the desired API type and call methods, for instance:
```

**Review by sfshaza2 — CHANGES_REQUESTED** [review]
Changing text to follow various conventions. Also, we prefer to define URLs below a section (or at the end of the page) to make the text more readable, but I'm going to let that pass.  :)

**Comment by kevmoo** [comment]
@Sfshaza – PTAL!

**Review by sfshaza2 — APPROVED** [review]
LGTM! And the build is happy! I'm happy to merge this, but I see you've also asked Andrew and Mit for a thumbs up, so I'll hold off.

**Comment by kevmoo** [comment]
It was more FYI for @RedBrogdon and @mit-mit – I need to get this landed so I can point my `go` link at SOMETHING!

**Comment by kevmoo** [comment]
...so please merge @Sfshaza 😁

**Pull request merged**